hub75/hub75.go

```diff
@@ -0,0 +1,292 @@
+	rowOffset         []uint32
+	buffer            [][]uint8 // [ColorDepth][(width * height * 3(rgb)) / 8]uint8
+	displayColor      uint16
+}
+
+// New returns a new HUB75 driver. Pass in a fully configured SPI bus.
+func New(b machine.SPI, latPin uint8, oePin uint8, aPin uint8, bPin uint8, cPin uint8, dPin uint8) Device {
+	pinA := machine.GPIO{aPin}
+	pinA.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})
+	pinB := machine.GPIO{bPin}
+	pinB.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})
+	pinC := machine.GPIO{cPin}
+	pinC.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})
+	pinD := machine.GPIO{dPin}
+	pinD.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})
+	pinOE := machine.GPIO{oePin}
+	pinOE.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})
+	pinLAT := machine.GPIO{latPin}
+	pinLAT.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})
+
+	return Device{
+		bus: b,
+		a:   pinA,
+		b:   pinB,
+		c:   pinC,
+		d:   pinD,
+		oe:  pinOE,
+		lat: pinLAT,
+	}
+}
+
+// Configure sets up the device.
+func (d *Device) Configure(cfg Config) {
+	if cfg.Width != 0 {
+		d.width = cfg.Width
+	} else {
+		d.width = 64
+	}
+	if cfg.Height != 0 {
+		d.height = cfg.Height
+	} else {
+		d.height = 32
+	}
+	if cfg.ColorDepth != 0 {
+		d.colorDepth = cfg.ColorDepth
+	} else {
+		d.colorDepth = 8
+	}
+	if cfg.RowPattern != 0 {
+		d.rowPattern = cfg.RowPattern
+	} else {
+		d.rowPattern = 16
+	}
+	if cfg.Brightness != 0 {
+		d.brightness = cfg.Brightness
+	} else {
+		d.brightness = 255
+	}
+
+	d.fastUpdate = cfg.FastUpdate
+	d.rowsPerBuffer = d.height / 2
+	d.panelWidth = 1
+	d.panelWidthBytes = (d.width / d.panelWidth) / 8
+	d.rowOffset = make([]uint32, d.height)
+	d.patternColorBytes = uint8((d.height / d.rowPattern) * (d.width / 8))
+	d.rowSetsPerBuffer = uint8(d.rowsPerBuffer / d.rowPattern)
+	d.sendBufferSize = uint16(d.patternColorBytes) * 3
+	d.colorStep = 256 / d.colorDepth
+	d.colorHalfStep = d.colorStep / 2
+	d.colorThirdStep = d.colorStep / 3
+	d.colorTwoThirdStep = 2 * d.colorThirdStep
+	d.buffer = make([][]uint8, d.colorDepth)
+	for i := range d.buffer {
+		d.buffer[i] = make([]uint8, (d.width*d.height*3)/8)
+	}
+
+	d.colorHalfStep = d.colorStep / 2
+	d.colorThirdStep = d.colorStep / 3
+	d.colorTwoThirdStep = 2 * d.colorThirdStep
+
+	d.a.Low()
+	d.b.Low()
+	d.c.Low()
+	d.d.Low()
+	d.oe.High()
+
+	var i uint32
+	for i = 0; i < uint32(d.height); i++ {
+		d.rowOffset[i] = (i%uint32(d.rowPattern))*uint32(d.sendBufferSize) + uint32(d.sendBufferSize) - 1
+	}
+}
+
+// SetPixelColor565 modifies the internal buffer in a single pixel.
+func (d *Device) SetPixelColor565(x int16, y int16, color uint16) {
+	r := uint8(((((color >> 11) & 0x1F) * 527) + 23) >> 6)
+	g := uint8(((((color >> 5) & 0x3F) * 259) + 33) >> 6)
+	b := uint8((((color & 0x1F) * 527) + 23) >> 6)
+	d.fillMatrixBuffer(x, y, r, g, b)
+}
+
+// SetPixel modifies the internal buffer in a single pixel.
+func (d *Device) SetPixel(x int16, y int16, c color.RGBA) {
```

Comment:
Should the signature for this be `func (d *Device) SetPixel(pt image.Point, c color.RGBA)`?